lib/aptible/cli/agent.rb

```diff
@@ -104,14 +117,38 @@ def login
           token_options[:expires_in] = duration
           token = Aptible::Auth::Token.create(token_options)
         rescue OAuth2::Error => e
-          if e.code == 'otp_token_required'
-            token_options[:otp_token] = options[:otp_token] ||
-                                        ask('2FA Token: ')
-            retry
+          if e.code != 'otp_token_required'
+            raise Thor::Error, 'Could not authenticate with given ' \
+                               "credentials: #{e.code}"
+          end
+
+          u2f = (e.response.parsed['exception_context'] || {})['u2f']
+
+          if u2f && security_key
```

Comment:
Most users, myself included, don't enter the OTP code until the prompt and trying to use the security key without the user specifying that they want to would change the default functionality. If we could run `u2f-auth` in parallel so the user can either touch the security key or enter a OTP code like on the website, that would be preferable. I'll look into that a little.

Reply:
I saw your commit adding the parallel, threaded code. I dropped a patch in a comment below to try to clarify this thought. I think it's fine to try only one and check each in turn and not parallelize it. If the user has U2F enabled but wants to use TOTP then they can use the command line flag to specify that.

What are your thoughts on that more minor change versus the threaded model?

Reply:
Since the TOTP isn't submitted until all of the information has been gathered, even if someone were to enter it as an argument, they'll probably want to enter their password in the prompt by which time the token may be invalid. That's why I always use the prompt.

Granted, with the addition of security key support, this would only be a PITA for anyone that has added a security key but isn't using it with the CLI for some reason. The only reasons I see that happening are if someone doesn't have their key on them (rare) or if they're using the CLI on a remote server. I'm not sure how many, if any, customers use the latter.

I like that the threaded model parallels the way we handle MFA when logging into the dashboard and I think our customers will appreciate it. That said, I have a tendency to get attached to my code so I'll defer to the team's judgement.

Reply:
Can you explain the role of the threads to me? 

My root concern is that threads always add some element of complexity or even non-determinism to the code and are best to be avoided when possible. I'm not sure that I'm grasping the value here.

It still seems like it'll execute in serial as the second thread blocks on `stdin`, right? There's only 1 terminal so you can't really enter both concurrently, and the second thread gets killed as soon as the first returns. This sets up something of a race condition, but the U2F one ought to win given its spawned first. But that's not guaranteed by the code. I've run this a few times to try to confirm my read and that seems to be the behavior.

This may just be my ruby inexperience showing? If so, I'm happy to jump on a call to talk through why the threads are needed here versus the plain sequential, single-threaded version. Or maybe someone with more ruby expertise like Ashley or Frank would need to take a look to sign-off. This one may just be over my head so feel free to bypass me for review and you can walk me through it after the fact.

Reply:
Ok, I guess I just needed fresh eyes on this.

One thread is waiting for `u2f-host` and the other is blocking on user input of the TOTP token. Whichever the user enters first pulls them through.

LGTM! Do you have a sense how those additional threads will interact with [u2f processes](https://github.com/aptible/aptible-cli/pull/253/files#diff-d16a05e7c9a26b8a4b2b1039d7bc0091R120) that is blocked on the `u2f-host` listening for devices? If the user enters their TOTP code causing the main CLI process to kill the u2f thread does that in turn invoke the [ensure](https://github.com/aptible/aptible-cli/pull/253/files#diff-d16a05e7c9a26b8a4b2b1039d7bc0091R129) block to clean up the resources?

Reply:
```ruby
puts 'start'

thread = Thread.new do
  begin
    puts 'begin'
    sleep
    puts 'shouldn\'t be seeing this'
  ensure
    puts 'ensure start'
    sleep 5
    puts 'ensure finish'
  end
end

puts 'waiting'
sleep 5

thread.kill.join

puts 'end'

# start
# waiting
# begin
# ensure start
# ensure finish
# end
```

Yes, when you `kill` a thread, the `ensure` block will run (assuming execution is still in the `begin` block) and `join` ensures that the thread exits before continuing.

This does still leave the chance that, if the user touches the key and enters their OTP at the same time, the thread will be killed while cleaning up. It looks like implementing `Thread.handle_interrupt` in the `ensure` block is the way to avoid this: https://ruby-doc.org/core-2.5.0/Thread.html#method-c-handle_interrupt

Reply:
`handle_interrupt` didn't work quite like I expected so I'm just checking if the thread is still waiting for input (`thread.status == 'sleep'`) and killing the thread if it is. Do you think this is sufficient?

Reply:
Honestly, I'm not familiar enough with Ruby threads to say if that's sufficient or not.

I guess my overall thought is this isn't critical functionality and it's isolated. The threaded version is a nicer UX and we have the code. If necessary, we can rip out the threaded implementation fairly easily since it's a standalone bit of code and still get the basic functionality in a single-threaded way. The overall risk here even if there's an edge case issue isn't grave.

Therefore, I'm okay with releasing this as is, with the understanding that if we do end up getting any odd behavior reports that we'll just fall back on to a single-threaded implementation and not sink a lot of time in hunting down bugs here. It can really hard to recreate/diagnose threading issues, especially across hardware/OS/ruby version differences so as long as we're all okay with keeping the maintenance burden threshold here low, then I'm 👍 on it.